pyro/ops/gaussian_sqrt.py

```diff
@@ -0,0 +1,385 @@
+import math
+
+import torch
+from torch.distributions.utils import lazy_property
+from torch.nn.functional import pad
+
+from pyro.distributions.util import broadcast_shape
+
+
+def triangularize(A, method='cholesky'):
+    """
+    Transforms a matrix to a lower triangular matrix such that A @ A.T = f(A) @ f(A).T
+
+    :param torch.Tensor A: input matrix
+    :param str method: either 'cholesky' or 'qr'
+    """
+    # TODO: benchmark two versions in real application, using 'qr' will be more stable but slower
+    if method == 'cholesky':
+        return A.matmul(A.transpose(-1, -2)).cholesky()
```

Comment:
@fritzo The operator `A.matmul(A.T)` will remove all the expanded rank at each step. Assume that `A` has shape `n x 4n` (which is the case when using `tensordot`), then `A.matmul(A.T)` will have shape `n x n`. I will add a test to guarantee that the square root's shape does not expanded in tensordot op.

Reply:
Similarly, QR of A.T where `A.shape = n x 4n` has shape `n x n`.

Reply:
But you're not currently triangularizing `Psqrt_a` in `.marginalize()`. I think to achieve bounded rank, you'll need to manually triangularize at the end of `.marginalize()`:
```py
if prec_sqrt.size(-1) > prec_sqrt.size(-2):
    prec_sqrt = triangularize(prec_sqrt)
```

Reply:
You are right! I missed this point. Let me see if there is an easy way to achieve this. :(